fvm/evm/handler/handler.go

```diff
@@ -0,0 +1,337 @@
+	"github.com/onflow/flow-go/fvm/errors"
+	"github.com/onflow/flow-go/fvm/evm/types"
+)
+
+// ContractHandler is responsible for triggering calls to emulator, metering,
+// event emission and updating the block
+//
+// TODO and Warning: currently database keeps a copy of roothash, and if after
+// commiting the changes by the evm we want to revert in this code we need to reset that
+// or we should always do all the checks and return before calling the emulator,
+// after that should be only event emissions and computation usage updates.
+// thats another reason we first check the computation limit before using.
+// in the future we might benefit from a view style of access to db passed as
+// a param to the emulator.
+type ContractHandler struct {
+	blockchain        types.BlockChain
+	backend           types.Backend
+	emulator          types.Emulator
+	lastExecutedBlock *types.Block
+	uuidIndex         uint64
+	totalSupply       uint64
+}
+
+var _ types.ContractHandler = &ContractHandler{}
+
+func NewContractHandler(
+	blockchain types.BlockChain,
+	backend types.Backend,
+	emulator types.Emulator,
+) *ContractHandler {
+	lastExecutedBlock, err := blockchain.LatestBlock()
+	// fatal error
+	if err != nil {
+		panic(err)
+	}
+
+	return &ContractHandler{
+		blockchain:        blockchain,
+		backend:           backend,
+		emulator:          emulator,
+		lastExecutedBlock: lastExecutedBlock,
+		uuidIndex:         lastExecutedBlock.UUIDIndex,
+		totalSupply:       lastExecutedBlock.TotalSupply,
+	}
+}
+
+// AllocateAddress allocates an address to be used by the bridged accounts
+func (h *ContractHandler) AllocateAddress() types.Address {
+	target := types.Address{}
+	// first 12 bytes would be zero
+	// the next 8 bytes would be incremented of uuid
+	binary.BigEndian.PutUint64(target[12:], h.lastExecutedBlock.UUIDIndex)
+	h.uuidIndex++
+
+	// TODO: if account exist try some new number
+	// if fe.State.Exist(target.ToCommon()) {
+	// }
+
+	h.updateLastExecutedBlock(h.lastExecutedBlock.StateRoot, gethTypes.EmptyRootHash)
+	return target
+}
+
+// AccountByAddress returns the account for the given address,
+// if isAuthorized is set, account is controlled by the FVM (bridged accounts)
+func (h *ContractHandler) AccountByAddress(addr types.Address, isAuthorized bool) types.Account {
+	return newAccount(h, addr, isAuthorized)
+}
+
+// LastExecutedBlock returns the last executed block
+func (h ContractHandler) LastExecutedBlock() *types.Block {
+	block, err := h.blockchain.LatestBlock()
+	handleError(err)
+	return block
+}
+
+func (h *ContractHandler) updateLastExecutedBlock(stateRoot, eventRoot gethCommon.Hash) {
+	h.lastExecutedBlock = types.NewBlock(
+		h.lastExecutedBlock.Height+1,
+		h.uuidIndex,
+		h.totalSupply,
+		stateRoot,
+		eventRoot,
+	)
+
+	err := h.blockchain.AppendBlock(h.lastExecutedBlock)
+	handleError(err)
+}
+
+// Run runs an rlpencoded evm transaction, collect the evm fees under the provided coinbase
```

Comment:
I would explain coinbase here, that this actually will return the fees back to the caller.

Reply:
updated the doc